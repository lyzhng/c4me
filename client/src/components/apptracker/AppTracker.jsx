import React from 'react';
import Axios from 'axios';
import AppTrackerItem from './AppTrackerItem';
import Scatterplot from './Scatterplot';

export default class AppTracker extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            students: [],
            minCollegeClass: 0,
            maxCollegeClass: Number.MAX_SAFE_INTEGER,
            highSchools: new Set(),
            currentHighSchool: '',
            appStatuses: [],
            scatterplot: false,
        };
    }

    setDefaultState = () => {
        this.setState({
            minCollegeClass: this.state.minCollegeClass === '' ? Number.MIN_SAFE_INTEGER : this.state.minCollegeClass,
            maxCollegeClass: this.state.maxCollegeClass === '' ? Number.MAX_SAFE_INTEGER : this.state.maxCollegeClass,
        });
    }

    handleChange = (e) => {
        this.setState({
            [e.target.name]: (isInt(e.target.value) ? +e.target.value : e.target.value)
        }, this.setDefaultState);
    }

    filter = (e) => {
        e.preventDefault();
        const students = this.state.students;
        for (const s of students) {
            const fitsCriteria = this.appFitsReq(s) && this.highSchoolFitsReq(s) && this.collegeClassFitsReq(s);
            // console.log('App Status Fits?', this.appFitsReq(s));
            // console.log('High School Fits?', this.highSchoolFitsReq(s));
            // console.log('College Class Fits?', this.collegeClassFitsReq(s));
            s.hidden = fitsCriteria ? false : true;
        }
        this.setState({ students: students });
    }

    collegeClassFitsReq = (student) => {
        const collegeClass = student.college_class;
        const { minCollegeClass, maxCollegeClass } = this.state;
        // console.log('College Class of Student', collegeClass);
        // console.log('Min College Class', minCollegeClass);
        // console.log('Max College Class', maxCollegeClass);
        return minCollegeClass <= collegeClass && collegeClass <= maxCollegeClass;
    }
    
    highSchoolFitsReq = (student) => {
        const highSchool = student.high_school_name;
        const { highSchools } = this.state;
        return highSchools.size === 0 || highSchools.has(highSchool); // might have to make it less specific
    }

    appFitsReq = (student) => {
        const { appStatuses } = this.state;
        const { applications } = student;
        const { name } = this.props.college;

        if (appStatuses.length === 0) {
            return true;
        } 

        for (const application of applications) {
            const { college, status } = application;
            const sameCollege = (application.college === name);
            const desired = appStatuses.includes(status);
            console.log(status, 'to', college);
            if (sameCollege && desired) {
                return true;
            }
        }
        return false;
    }

    addHighSchool = () => {
        this.setState({ highSchools: new Set([...this.state.highSchools, this.state.currentHighSchool]) });
    }

    removeHighSchool = (e) => {
        e.preventDefault();
        const { highSchools } = this.state;
        const name = document.querySelector('.high-school-unit > .high-school-name').textContent;
        highSchools.delete(name);
        this.setState({ highSchools });
    }

    handleAppStatuses = (e) => {
        const choice = e.target.name;
        if (this.state.appStatuses.includes(choice)) {
            let filteredCopy = this.state.appStatuses.filter((status) => status !== choice);
            this.setState({ appStatuses: filteredCopy });
        } else {
            this.setState({ appStatuses: [...this.state.appStatuses, choice] });
        }
    }

    async componentDidMount() {
        if (this.props.college) {
            const collegeName = this.props.college.name;
            const resp = await Axios.post('/retrievestudents', { query: collegeName });
            this.setState({ students: resp.data.students });
            console.log('Students:', this.state.students);
        }
    }

    toggleScatterplot = () => {
        this.setState({ scatterplot: !this.state.scatterplot });
    }

    render() {
        return (
            <div>
                <div className="filters">
                    <h2>App Tracker for {this.props.college.name}</h2>
                    {/* college class */}
                    <div className="row">
                        <label htmlFor="minCollegeClass">Min. College Class</label>
                        <input type="number" name="minCollegeClass" min="0" onChange={this.handleChange} />
                    </div>
                    <div className="row">
                        <label htmlFor="maxCollegeClass">Max. College Class</label>
                        <input type="number" name="maxCollegeClass" id="" min="0" onChange={this.handleChange} />
                    </div>
                    {/* high schools */}
                    <div className="row">
                        <label htmlFor="currentHighSchool">High Schools</label>
                        <input type="text" name="currentHighSchool" id="" value={this.state.currentHighSchool} onChange={this.handleChange} className="current-hs"/>
                        <button type="submit" onClick={this.addHighSchool}
                            disabled={!this.state.currentHighSchool.match(/^([\w\-\,\.]+\s*)+$/g)} className="btn-add-hs">Add High School</button>
                    </div>
                    <div className="high-school-list">
                        {
                            [...this.state.highSchools].map((highSchool) => {
                                return (
                                    <div key={highSchool} className="high-school-unit">
                                        <h5 className="high-school-name">{highSchool}</h5>
                                        <button onClick={this.removeHighSchool}>Remove High School</button>
                                    </div>
                                );
                            })
                        }
                    </div>
                    <div className="row">
                        <label htmlFor="appStatuses">Application Statuses:</label>
                        <input type="checkbox" name="pending" id="" onChange={this.handleAppStatuses} />
                        <label htmlFor="pending">Pending</label>
                        <input type="checkbox" name="accepted" id="" onChange={this.handleAppStatuses} />
                        <label htmlFor="accepted">Accepted</label>
                        <input type="checkbox" name="denied" id="" onChange={this.handleAppStatuses} />
                        <label htmlFor="denied">Denied</label>
                        <input type="checkbox" name="deferred" id="" onChange={this.handleAppStatuses} />
                        <label htmlFor="deferred">Deferred</label>
                        <input type="checkbox" name="wait-listed" id="" onChange={this.handleAppStatuses} />
                        <label htmlFor="wait-listed">Wait-listed</label>
                        <input type="checkbox" name="withdrawn" id="" onChange={this.handleAppStatuses} />
                        <label htmlFor="withdrawn">Withdrawn</label>
                    </div>
                    <button onClick={this.filter} className="btn-filter">Apply Filters</button>
                    <label htmlFor="scatterplot">Scatterplot: </label>
                    <input type="checkbox" name="scatterplot" id="" onChange={this.toggleScatterplot} checked={this.state.scatterplot} />
                </div>
                {
                    this.state.scatterplot ?
                        <Scatterplot students={this.state.students} college={this.props.college}/> :
                    <div className="student-list">
                        {this.state.students.map((student) => <AppTrackerItem key={student._id} student={student} />)}
                    </div>
                }
            </div>
        );
    }
}

function isInt(n) {  //copied from https://stackoverflow.com/questions/5630123/javascript-string-integer-comparisons
    return /^[+-]?\d+$/.test(n);
};