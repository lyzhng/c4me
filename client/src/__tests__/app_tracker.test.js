import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import AppTracker from '../components/apptracker/AppTracker';

Enzyme.configure({ adapter: new Adapter() });

let appCount = 0;
const students = [];

const sampleProps = {
  college: {
    name: 'Stony Brook University'
  }
}

describe('AppTracker Component', () => {
  beforeAll(() => {
    createSampleStudents();
  });

  it('should have initial states', () => {
    const wrapper = shallow(<AppTracker {...sampleProps} />);
    const {
      students,
      minCollegeClass,
      maxCollegeClass,
      highSchools,
      currentHighSchool,
      appStatuses,
    } = wrapper.state();
    expect(students.length).toBe(0);
    expect(minCollegeClass).toBe(0);
    expect(maxCollegeClass).toBe(Number.MAX_SAFE_INTEGER);
    expect(highSchools.size).toBe(0);
    expect(currentHighSchool).toBe('');
    expect(appStatuses.length).toBe(0);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should add the high school to the high school state', function() {
    const wrapper = shallow(<AppTracker {...sampleProps} />);
    const sampleHighSchool = 'Ward Melville Senior High School';
    wrapper.setState({ highSchools: new Set() });
    wrapper.setState({ currentHighSchool: sampleHighSchool });
    wrapper.find('button.btn-add-hs').simulate('click');
    const { highSchools } = wrapper.state();
    expect(highSchools.size).toBe(1);
    expect(highSchools.has(sampleHighSchool)).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should return true if high school state is empty', function () {
    const wrapper = shallow(<AppTracker {...sampleProps} />);
    wrapper.setState({ students });
    wrapper.setState({ highSchools: new Set() });
    for (const s of students) { 
      wrapper.update();
      expect(wrapper.instance().highSchoolFitsReq(s)).toBe(true);
    }
  });

  it('should return true if the high school fits the filter requirements', function() {
    const wrapper = shallow(<AppTracker {...sampleProps} />);
    wrapper.setState({ students });
    const hs = new Set();
    hs.add('Ward Melville Senior High School');
    wrapper.setState({ highSchools: hs });
    for (const s of students) {
      wrapper.update();
      if (hs.has(s.high_school_name)) {
        console.log(s);
        expect(wrapper.instance().highSchoolFitsReq(s)).toBe(true);
      } else {
        expect(wrapper.instance().highSchoolFitsReq(s)).toBe(false);
      }
    }
  });

  it('should return true if student.college_class is within bounds', function() {
    const wrapper = shallow(<AppTracker {...sampleProps} />);
    wrapper.setState({ students });
    wrapper.setState({ minCollegeClass: 2020 });
    wrapper.setState({ maxCollegeClass: 2025 });
    const {
      minCollegeClass,
      maxCollegeClass
    } = wrapper.state();
    for (const s of students) {
      const withinBounds = minCollegeClass <= s.college_class && s.college_class <= maxCollegeClass;
      if (withinBounds) {
        expect(wrapper.instance().collegeClassFitsReq(s)).toBe(true);
      } else {
        expect(wrapper.instance().collegeClassFitsReq(s)).toBe(false);
      }
      wrapper.update();
    }
  });

  it('should return true if student application to that college is filtered', function() {
    const wrapper = shallow(<AppTracker {...sampleProps} />);
    wrapper.setState({ students });
    wrapper.setState({
      appStatuses: [
        'accepted',
      ]
    });
    const {
      appStatuses,
    } = wrapper.state();

    if (appStatuses.length === 0) {
      expect(wrapper.instance().appFitsReq(s)).toBe(true);
    }

    for (const s of students) {
      const { status: appStatus, college } = s.applications[0];
      if (appStatuses.includes(appStatus) && college === sampleProps.college.name) {
          expect(wrapper.instance().appFitsReq(s)).toBe(true);
      } else {
          expect(wrapper.instance().appFitsReq(s)).toBe(false);
      }
      wrapper.update();
    }
  });

  it('should filter the correct students', () => {
    const wrapper = shallow(<AppTracker {...sampleProps} />);
    wrapper.setState({ students });
    const {
      collegeClassFitsReq,
      highSchoolFitsReq,
      appFitsReq,
    } = wrapper.instance();
    const hs = new Set();
    hs.add('Ward Melville Senior High School');
    wrapper.setState({ minCollegeClass: 2020 });
    wrapper.setState({ maxCollegeClass: 2025 });
    wrapper.setState({ highSchools: hs });
    wrapper.setState({ appStatuses: ['accepted'] });
    wrapper.find('button.btn-filter').simulate('click');
    wrapper.update();
    for (const s of students) {
      if (collegeClassFitsReq(s) && highSchoolFitsReq(s) && appFitsReq(s)) {
        expect(s.hidden).toBe(false);
      } else {
        expect(s.hidden).toBe(true);
      }
    }
  });
});

function createStudent(userid, password, high_school_name, college_class, college, status) {
  const student = {
    _id: userid,
    userid,
    password,
    high_school_name,
    college_class,
    applications: [
      createApplication(appCount++, 'Stony Brook University', 'accepted'),
    ]
  };
  return student;
}

function createApplication(_id, college, status) {
  const app = {
    _id,
    college,
    status
  };
  return app;
}

function createSampleStudents() {
  const student1 = createStudent('sample1', '12345678', 'Ward Melville Senior High School', 2024, 'Stony Brook University', 'accepted');
  const student2 = createStudent('sample2', '12345678', 'Francis Lewis High School', 2021, 'Princeton University', 'deferred');
  const student3 = createStudent('sample3', '12345678', 'Ward Melville Senior High School', 2030, 'Princeton University', 'wait-listed');
  const student4 = createStudent('sample4', '12345678', 'Brooklyn Technical High School', 2027, 'Princeton University', 'accepted');
  const student5 = createStudent('sample5', '12345678', 'Stuyvesant High School', 2027, 'Harvard University', 'accepted');
  students.push(student1);
  students.push(student2);
  students.push(student3);
  students.push(student4);
  students.push(student5);
}