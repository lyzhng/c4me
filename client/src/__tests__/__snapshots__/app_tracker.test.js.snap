// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppTracker Component should add the high school to the high school state 1`] = `
<div>
  <div
    className="filters"
  >
    <h2>
      App Tracker for 
      Stony Brook University
    </h2>
    <div
      className="row"
    >
      <label
        htmlFor="minCollegeClass"
      >
        Min. College Class
      </label>
      <input
        min="0"
        name="minCollegeClass"
        onChange={[Function]}
        type="number"
      />
    </div>
    <div
      className="row"
    >
      <label
        htmlFor="maxCollegeClass"
      >
        Max. College Class
      </label>
      <input
        id=""
        min="0"
        name="maxCollegeClass"
        onChange={[Function]}
        type="number"
      />
    </div>
    <div
      className="row"
    >
      <label
        htmlFor="currentHighSchool"
      >
        High Schools
      </label>
      <input
        className="current-hs"
        id=""
        name="currentHighSchool"
        onChange={[Function]}
        type="text"
        value="Ward Melville Senior High School"
      />
      <button
        className="btn-add-hs"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Add High School
      </button>
    </div>
    <div
      className="high-school-list"
    >
      <div
        className="high-school-unit"
        key="Ward Melville Senior High School"
      >
        <h5
          className="high-school-name"
        >
          Ward Melville Senior High School
        </h5>
        <button
          onClick={[Function]}
        >
          Remove High School
        </button>
      </div>
    </div>
    <div
      className="row"
    >
      <label
        htmlFor="appStatuses"
      >
        Application Statuses:
      </label>
      <input
        id=""
        name="pending"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="pending"
      >
        Pending
      </label>
      <input
        id=""
        name="accepted"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="accepted"
      >
        Accepted
      </label>
      <input
        id=""
        name="denied"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="denied"
      >
        Denied
      </label>
      <input
        id=""
        name="deferred"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="deferred"
      >
        Deferred
      </label>
      <input
        id=""
        name="wait-listed"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="wait-listed"
      >
        Wait-listed
      </label>
      <input
        id=""
        name="withdrawn"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="withdrawn"
      >
        Withdrawn
      </label>
    </div>
    <button
      className="btn-filter"
      onClick={[Function]}
    >
      Apply Filters
    </button>
    <label
      htmlFor="scatterplot"
    >
      Scatterplot: 
    </label>
    <input
      checked={false}
      id=""
      name="scatterplot"
      onChange={[Function]}
      type="checkbox"
    />
  </div>
  <div
    className="student-list"
  />
</div>
`;

exports[`AppTracker Component should have initial states 1`] = `
<div>
  <div
    className="filters"
  >
    <h2>
      App Tracker for 
      Stony Brook University
    </h2>
    <div
      className="row"
    >
      <label
        htmlFor="minCollegeClass"
      >
        Min. College Class
      </label>
      <input
        min="0"
        name="minCollegeClass"
        onChange={[Function]}
        type="number"
      />
    </div>
    <div
      className="row"
    >
      <label
        htmlFor="maxCollegeClass"
      >
        Max. College Class
      </label>
      <input
        id=""
        min="0"
        name="maxCollegeClass"
        onChange={[Function]}
        type="number"
      />
    </div>
    <div
      className="row"
    >
      <label
        htmlFor="currentHighSchool"
      >
        High Schools
      </label>
      <input
        className="current-hs"
        id=""
        name="currentHighSchool"
        onChange={[Function]}
        type="text"
        value=""
      />
      <button
        className="btn-add-hs"
        disabled={true}
        onClick={[Function]}
        type="submit"
      >
        Add High School
      </button>
    </div>
    <div
      className="high-school-list"
    />
    <div
      className="row"
    >
      <label
        htmlFor="appStatuses"
      >
        Application Statuses:
      </label>
      <input
        id=""
        name="pending"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="pending"
      >
        Pending
      </label>
      <input
        id=""
        name="accepted"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="accepted"
      >
        Accepted
      </label>
      <input
        id=""
        name="denied"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="denied"
      >
        Denied
      </label>
      <input
        id=""
        name="deferred"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="deferred"
      >
        Deferred
      </label>
      <input
        id=""
        name="wait-listed"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="wait-listed"
      >
        Wait-listed
      </label>
      <input
        id=""
        name="withdrawn"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="withdrawn"
      >
        Withdrawn
      </label>
    </div>
    <button
      className="btn-filter"
      onClick={[Function]}
    >
      Apply Filters
    </button>
    <label
      htmlFor="scatterplot"
    >
      Scatterplot: 
    </label>
    <input
      checked={false}
      id=""
      name="scatterplot"
      onChange={[Function]}
      type="checkbox"
    />
  </div>
  <div
    className="student-list"
  />
</div>
`;
